### PR TITLE
IGNITE-15650 : Statistics for platform services. V2 : Static version.

### PR DESCRIPTION
**Pull request by Vladsz83**
This PR brings service statistics for the platform services (.NET). Since we don't know what is the platform service being deployed, we have to provide extended info of platform service to build proper statistics. So, PlatformServiceConfiguration has been involved. Also, LazyServiceConfiguration got extended.

There is 'on demand' version: #9769

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/platform/services/PlatformServices.java [hunk 1]
```diff
@@ -822,5 +826,7 @@ private static void writeFailedConfiguration(BinaryRawWriterEx w, ServiceConfigu
         if (svcCfg.getNodeFilter() instanceof PlatformClusterNodeFilterImpl)
             dotnetFilter = ((PlatformClusterNodeFilterImpl)svcCfg.getNodeFilter()).getInternalPredicate();
         w.writeObjectDetached(dotnetFilter);
+
+        w.writeBoolean(svcCfg.isStatisticsEnabled());
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/platform/services/PlatformServices.java:830]: Should we write known methods here?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java [hunk 2]
```diff
@@ -638,7 +642,7 @@ private void ensure(boolean cond, String desc, @Nullable Object v) {
      */
     private PreparedConfigurations<IgniteUuid> prepareServiceConfigurations(Collection<ServiceConfiguration> cfgs,
         IgnitePredicate<ClusterNode> dfltNodeFilter) {
-        List<ServiceConfiguration> cfgsCp = new ArrayList<>(cfgs.size());
+        List<LazyServiceConfiguration> cfgsCp = new ArrayList<>(cfgs.size());
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java:645]: In most cases there is no need to change `ServiceConfiguration` to `LazyServiceConfiguration`. The latter inherits from the former, so we can safely use `ServiceConfiguration` everywhere and cast it to `Lazy...` in single place only.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/service/LazyServiceConfiguration.java [hunk 3]
```diff
@@ -40,6 +40,10 @@
     /** */
     private byte[] srvcBytes;
 
+    /** Known service methods to build service statistics. */
+    @GridToStringExclude
+    private String[] knownMtdNames;
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/service/LazyServiceConfiguration.java:45]: It used only for platform services and then this field is very misleading. I propose to store those methods inside of `PlatformService`. Is it possible?

File: modules/platforms/dotnet/Apache.Ignite.Core.Tests/Services/ServicesAsyncWrapper.cs [hunk 4]
```diff
@@ -170,6 +170,7 @@ public ICollection<IServiceDescriptor> GetServiceDescriptors()
         }
 
         /** <inheritDoc /> */
+#pragma warning disable 618
```
Comment (timoninmaxim) [inline modules/platforms/dotnet/Apache.Ignite.Core.Tests/Services/ServicesAsyncWrapper.cs:173]: I'm not sure, but this `pragma` is the single change in the file, do we actually need it?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/platform/services/PlatformServices.java (hunk 1 above)
Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/platform/services/PlatformServices.java:830]: Back to .NET? Not sure. The platform doesn't need to know the methods were tried to get covered by the statistics. On the platform side the methods are known.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java (hunk 2 above)
Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java:645]: Fixed

File: modules/core/src/main/java/org/apache/ignite/internal/processors/service/LazyServiceConfiguration.java (hunk 3 above)
Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/service/LazyServiceConfiguration.java:45]: PlatformService, like any service, has many instances. ServiceCOnfig or service context has single instance. It's better to store one instance of shared resource/config.

File: modules/platforms/dotnet/Apache.Ignite.Core.Tests/Services/ServicesAsyncWrapper.cs (hunk 4 above)
Comment (Vladsz83) [reply modules/platforms/dotnet/Apache.Ignite.Core.Tests/Services/ServicesAsyncWrapper.cs:173]: Fixed

File: modules/core/src/test/java/org/apache/ignite/platform/PlatformDeployServiceTask.java [hunk 5]
```diff
@@ -671,10 +681,101 @@ public void sleep(long delayMs) {
...
+
+                    if (res.getException() != null) {
+                        throw new IgniteException("Unable to count invocations in service metrics. Job failed on " +
+                            "node [" + res.getNode() + "]: " + res.getException().getMessage(), res.getException());
+                    }
+
+                    if (res.getData() == null)
+                        continue;
+
+                    cnt += (Long)res.getData();
```
Comment (timoninmaxim) [inline modules/core/src/test/java/org/apache/ignite/platform/PlatformDeployServiceTask.java:722]: Obsolete cast to `Long`.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/platform/services/PlatformServiceConfiguration.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+     * @return Known method names of platform service.
+     */
+    public String[] mtdNames() {
+        return mtdNames;
+    }
+
+    /**
+     * Sets known method names of platform service.
+     */
+    void setMtdNames(String[] mtdNames) {
```
[the remark this answers is not in the document]
Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/platform/services/PlatformServiceConfiguration.java:49]: Fixed

File: modules/core/src/main/java/org/apache/ignite/internal/processors/platform/services/PlatformServiceConfiguration.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+        this.mtdNames = mtdNames == null ? new String[0] : mtdNames;
```
[the remark this answers is not in the document]
Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/platform/services/PlatformServiceConfiguration.java:50]: Fixed

File: modules/platforms/dotnet/Apache.Ignite.Core/Services/ServiceConfiguration.cs [hunk 8]
```diff
@@ -83,6 +91,29 @@ internal void Write(IBinaryRawWriter w)
...
             else
                 w.WriteObject<object>(null);
+
+            w.WriteBoolean(StatisticsEnabled);
+
+            WriteExtraDescription(w);
+        }
+
+        /// <summary>
+        /// Provides extra intel about platform service to avoid on-demand creation of service statistics on any
```
[the remark this answers is not in the document]
Comment (Vladsz83) [reply modules/platforms/dotnet/Apache.Ignite.Core/Services/ServiceConfiguration.cs:101]: Fixed

File: modules/core/src/test/java/org/apache/ignite/platform/PlatformDeployServiceTask.java [hunk 9]
```diff
@@ -671,10 +681,101 @@ public void sleep(long delayMs) {
             }
         }
 
+        /**
+         * Calculates number of registered values among the service statistics.
+         *
+         * @return Number of registered values among the service statistics. Or {@code null} if no statistics found
```
[the remark this answers is not in the document]
Comment (Vladsz83) [reply modules/core/src/test/java/org/apache/ignite/platform/PlatformDeployServiceTask.java:687]: Fixed

File: modules/core/src/test/java/org/apache/ignite/platform/PlatformDeployServiceTask.java (hunk 5 above)
Comment (Vladsz83) [reply modules/core/src/test/java/org/apache/ignite/platform/PlatformDeployServiceTask.java:722]: Nope. The result has no certain type. Idea says obsolete, but not. You might try to remove the cast.

Comment (timoninmaxim) [reply modules/core/src/test/java/org/apache/ignite/platform/PlatformDeployServiceTask.java:721]: Yes, but why `Long`? The method `reduce` returns Integer, why do we operate with long here? Should we return Long then?

File: modules/platforms/dotnet/Apache.Ignite.Core.Tests/Services/ServicesTest.cs [hunk 10]
```diff
@@ -1330,6 +1411,91 @@ private void DoTestBinary(IJavaService svc, IJavaService binSvc, bool isPlatform
                     .GetField<int>("Field"));
         }
 
+        /// <summary>
+        /// Tests platform service statistics.
+        /// </summary>
+        private void DoTestMetrics(IServices producer, IServices consumer, IServiceCallContext callCtx, bool dyn)
```
[the remark this answers is not in the document]
Comment (timoninmaxim) [reply modules/platforms/dotnet/Apache.Ignite.Core.Tests/Services/ServicesTest.cs:1417]: As far as I understood a pure Java service is the `helperSvc`. Could we check metrics of `helperSvc.testNumberOfInvocations` in this method too? Just to be sure that pure java service will work correctly. Looks like it just adding a single assert.

File: modules/core/src/test/java/org/apache/ignite/platform/PlatformDeployServiceTask.java (hunk 5 above)
Comment (Vladsz83) [reply modules/core/src/test/java/org/apache/ignite/platform/PlatformDeployServiceTask.java:721]: Turned to int

File: modules/platforms/dotnet/Apache.Ignite.Core.Tests/Services/ServicesTest.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (Vladsz83) [reply modules/platforms/dotnet/Apache.Ignite.Core.Tests/Services/ServicesTest.cs:1417]: Hm. Yep. But we have tests for Java service metrics. Why do we need it on .NET platform services for not-platform services?

[the remark this answers is not in the document]
Comment (Vladsz83) [reply modules/platforms/dotnet/Apache.Ignite.Core.Tests/Services/ServicesTest.cs:1434]: As an example, I added TestJavaServiceStatistics(). I check invocations by itself. I don't think we need to check java service metrics from platform services.

**Review by timoninmaxim — APPROVED** [review]

**Pull request merged**